blockprod/src/detail/builder.rs

```diff
@@ -0,0 +1,220 @@
+        self.block_makers_tx.send(BlockMakerControlCommand::StopBecauseNewTip(
+            new_tip_id,
+            new_tip_height,
+        )).expect("The channel can never be disconnected since there's a receiver always alive in self");
+        Ok(())
+    }
+
+    pub fn stop_all_block_makers(&self) -> Result<(), BlockProductionError> {
+        self.block_makers_tx.send(BlockMakerControlCommand::JustStop).expect(
+            "The channel can never be disconnected since there's a receiver always alive in self",
+        );
+        Ok(())
+    }
+
+    pub async fn trigger_new_block_production(
+        &self,
+        current_tip_id: Id<Block>,
+        current_tip_height: BlockHeight,
+    ) -> Result<(), BlockProductionError> {
+        if !self.enabled {
+            return Ok(());
+        }
+
+        let chain_config = self.chain_config.clone();
+        let chainstate_handle = self.chainstate_handle.clone();
+        let mempool_handle = self.mempool_handle.clone();
+        let time_getter = self.time_getter.clone();
+        let command_receiver = self.block_maker_rx.clone();
+        tokio::spawn(async move {
+            BlockMaker::new(
+                chain_config,
+                chainstate_handle,
+                mempool_handle,
+                time_getter,
+                current_tip_id,
+                current_tip_height,
+                command_receiver,
+            )
+            .run()
+            .await
+        });
+        Ok(())
+    }
+
+    pub async fn run(&mut self) -> Result<(), BlockProductionError> {
+        let mut mempool_rx = self.subscribe_to_mempool_events().await?;
+        let mut chainstate_rx = self.subscribe_to_chainstate_events().await?;
+
+        loop {
+            tokio::select! {
+                // when we receive information from chainstate that we have a new tip, we stop building the current block,
+                // and expect the mempool to soon send a trigger for a new command
+                block_info = chainstate_rx.recv().fuse() => {
+                    let (block_id, block_height) = block_info.ok_or(BlockProductionError::ChainstateChannelClosed)?;
+                    self.stop_building(block_id, block_height)?;
+                }
+                block_info = mempool_rx.recv().fuse() => {
+                    let (block_id, block_height) = block_info.ok_or(BlockProductionError::MempoolChannelClosed)?;
+                    self.trigger_new_block_production(block_id, block_height).await?;
+                }
+                event = self.builder_rx.recv().fuse() => match event.ok_or(BlockProductionError::BlockBuilderChannelClosed)? {
+                    BlockBuilderControlCommand::Stop => {
+                        self.enabled = false;
+                        self.stop_all_block_makers()?;
+                    },
+                    BlockBuilderControlCommand::Start => {
+                        self.enabled = true;
+                    },
```

Comment:
It's a convention. 

Stopping must stop everything because there's no point in continuing if we want to stop. While when we start, we may be too late to the game of staking/mining the current block. So we just wait for the next tip to start.